src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -54,366 +57,1142 @@ internal static class ThreadPoolGlobals
         {
-            private static volatile WorkStealingQueue[] _queues = new WorkStealingQueue[0];
+            /// <summary>The current enqueue segment.</summary>
+            internal GlobalQueueSegment _enqSegment;
+            /// <summary>The current dequeue segment.</summary>
+            internal GlobalQueueSegment _deqSegment;
 
-            public static WorkStealingQueue[] Queues => _queues;
+            /// <summary>
+            /// Initializes a new instance of the <see cref="GlobalQueue"/> class.
+            /// </summary>
+            internal GlobalQueue()
+            {
+                _addSegmentLock = new object();
+                _enqSegment = _deqSegment = new GlobalQueueSegment(InitialSegmentLength);
+            }
 
-            public static void Add(WorkStealingQueue queue)
+            // for debugging
+            internal int Count
             {
-                Debug.Assert(queue != null);
-                while (true)
+                get
                 {
-                    WorkStealingQueue[] oldQueues = _queues;
-                    Debug.Assert(Array.IndexOf(oldQueues, queue) == -1);
-
-                    var newQueues = new WorkStealingQueue[oldQueues.Length + 1];
-                    Array.Copy(oldQueues, 0, newQueues, 0, oldQueues.Length);
-                    newQueues[newQueues.Length - 1] = queue;
-                    if (Interlocked.CompareExchange(ref _queues, newQueues, oldQueues) == oldQueues)
+                    int count = 0;
+                    for (var s = _deqSegment; s != null; s = s._nextSegment)
                     {
-                        break;
+                        count += s.Count;
                     }
+                    return count;
+                }
+            }
+
+            /// <summary>
+            /// Adds an object to the top of the queue
+            /// </summary>
+            internal void Enqueue(object item)
+            {
+                // try enqueuing. Should normally succeed unless we need a new segment.
+                if (!_enqSegment.TryEnqueue(item))
+                {
+                    // If we're unable to enque, this segment will never take enqueues again.
+                    // we need to take a slow path that will try adding a new segment.
+                    EnqueueSlow(item);
                 }
             }
 
-            public static void Remove(WorkStealingQueue queue)
+            /// <summary>
+            /// Slow path for enqueue, adding a new segment if necessary.
+            /// </summary>
+            private void EnqueueSlow(object item)
             {
-                Debug.Assert(queue != null);
-                while (true)
+                for (; ; )
                 {
-                    WorkStealingQueue[] oldQueues = _queues;
-                    if (oldQueues.Length == 0)
+                    GlobalQueueSegment currentSegment = _enqSegment;
+                    if (currentSegment.TryEnqueue(item))
```

Comment:
Copied form ConcurrentQueue. I think it is just to make it easier to read. This method should be rarely called since every time the buffer size is doubled.

Will try moving to the bottom of the loop

---
In reply to: [250368421](https://github.com/dotnet/coreclr/pull/18403#discussion_r250368421) [](ancestors = 250368421)

Reply:
I suppose the rationale could be that the field is re-read and we only want to freeze that segment if the fast path fails. Alternatively, the value for the field that was read in the fast path could be passed in to retain that rationale while still not trying the fast path before taking the lock, and as you said to move the call to `TryEnqueue` to the bottom of the loop.

Reply:
Yes that is correct. It would require capturing the segment in a local and passing as an extra argument if we want to double the segment size.
 I.E. is results in making fast path noticeably more complex (since it is very simple) for the purpose of simplifying slow path that compared to the fast path is nearly never taken.
I think it makes sense to keep this as-is.

---
In reply to: [250443610](https://github.com/dotnet/coreclr/pull/18403#discussion_r250443610) [](ancestors = 250443610)